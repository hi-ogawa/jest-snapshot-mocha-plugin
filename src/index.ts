import * as assert from "assert";
import type { Context } from "mocha";
import { SnapshotState } from "jest-snapshot";
import type { SnapshotStateType } from "jest-snapshot";
import type {
  SnapshotStateOptions,
  SnapshotMatchOptions,
} from "jest-snapshot/build/State";
import * as path from "path";

const NAME = "jest-snapshot-mocha-plugin";
const SNAPSHOT_EXTENSION = ".snap";

export type ResolveSnapshotPath = (
  testPath: string,
  snapshotExt: string
) => string;

export interface PluginConfig {
  updateSnapshot: boolean;
  resolveSnapshotPath?: ResolveSnapshotPath;
}

export interface PluginState {
  title: string;
  snapshotPath: string;
  snapshotState: SnapshotStateType;
}

export class SnapshotError extends assert.AssertionError {
  constructor(
    result: ReturnType<SnapshotStateType["match"]>,
    snapshotPath: string,
    stackStartFn: Function
  ) {
    super({
      actual: result.actual,
      expected: result.expected,
      operator: "matchSnapshot",
      stackStartFn,
    });

    // Snapshot not found
    if (result.expected === undefined) {
      this.message = `Snapshot not found (${snapshotPath})`;
      return;
    }

    // Generate message via assert equal
    // (Unfortunately jest-snapshot's `printSnapshotAndReceived` is not exported ...)
    try {
      assert.strict.equal(result.actual, result.expected);
    } catch (e: unknown) {
      if (e instanceof Error) {
        // Swap 1st line
        this.message = [
          `Snapshot not match (${snapshotPath})`,
          e.message.split("\n").slice(1),
        ].join("\n");
      }
    }
  }
}

type MochaHooks = {
  beforeEach: (this: Context) => void;
  afterEach: (this: Context) => void;
};

export class JestSnapshotMochaPlugin {
  private state: PluginState | undefined;
  private resolveSnapshotPath: ResolveSnapshotPath;

  constructor(private config: PluginConfig) {
    this.resolveSnapshotPath =
      config.resolveSnapshotPath ?? defaultResolveSnapshotPath;
  }

  matchSnapshot = (actual: unknown): void => {
    if (!this.state) {
      throw new Error(`[${NAME}] no state (possibly due to misconfigurationn)`);
    }
    const { title, snapshotPath, snapshotState } = this.state;

    const matchOptions: SnapshotMatchOptions = {
      received: actual,
      testName: title,
      isInline: false,
    };
    const result = snapshotState.match(matchOptions);
    if (result.pass) return;
    throw new SnapshotError(result, snapshotPath, this.matchSnapshot);
  };

  beforeEach(context: Context): void {
    const test = context.currentTest;
    if (!test) {
      log("no context.currentTest");
      return;
    }

    const file = test.file;
    if (!file) {
      log("no context.currentTest.file");
      return;
    }

    const title = test.fullTitle();
    const snapshotPath = this.resolveSnapshotPath(file, SNAPSHOT_EXTENSION);
    const snapshotOptions: SnapshotStateOptions = {
      updateSnapshot: this.config.updateSnapshot ? "all" : "none",
      // Other flags have been changing between versions,
      // but those are related to inline snapshot feature,
      // so the older verions of jest-snapshot should still work.
      prettierPath: "",
      snapshotFormat: {},
    };
    const snapshotState = new SnapshotState(snapshotPath, snapshotOptions);
    this.state = { title, snapshotPath, snapshotState };
  }

  afterEach(): void {
    if (!this.state) return;

    const { snapshotState } = this.state;
    this.state = undefined;

    // TODO: Does it partially update (e.g. when grepping with -g)?
    snapshotState.save();
  }

  get mochaHooks(): MochaHooks {
    const self = this;
    return {
      beforeEach: function (this: Context) {
        self.beforeEach(this);
      },
      afterEach: function (this: Context) {
        self.afterEach();
      },
    };
  }
}

function defaultResolveSnapshotPath(
  testPath: string,
  snapshotExt: string
): string {
  const testDirname = path.dirname(path.relative(".", testPath));
  const testFileNoExt = path
    .basename(testPath)
    .split(".")
    .slice(0, -1)
    .join(".");
  return path.join(testDirname, "__snapshots__", testFileNoExt + snapshotExt);
}

function log(...args: any): void {
  console.error(`[${NAME}]`, ...args);
}
