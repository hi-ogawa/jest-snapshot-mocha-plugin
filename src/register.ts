//
// Default 'register' script
//
import { JestSnapshotMochaPlugin } from "./index";
import * as process from "process";

export const updateSnapshot: boolean = !!process.env.UPDATE_SNAPSHOT;
export const plugin = new JestSnapshotMochaPlugin({ updateSnapshot });
export const matchSnapshot = plugin.matchSnapshot;

// Interface for --require https://mochajs.org/#defining-a-root-hook-plugin
export const mochaHooks = plugin.mochaHooks;
