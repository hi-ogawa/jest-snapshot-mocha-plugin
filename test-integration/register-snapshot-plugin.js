const process = require("process");
const { JestSnapshotMochaPlugin } = require("../build/tsc/src");

const plugin = new JestSnapshotMochaPlugin({
  updateSnapshot: !!process.env.UPDATE_SNAPSHOT,
});

global.matchSnapshot = plugin.matchSnapshot;
exports.mochaHooks = plugin.mochaHooks;
