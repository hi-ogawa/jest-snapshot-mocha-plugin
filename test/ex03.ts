import * as assert from "assert/strict";
import { describe, it } from "mocha";
import { matchSnapshot, updateSnapshot } from "./register-snapshot-plugin";
import { SnapshotError } from "../src";
import * as fs from "fs";

const ACTUAL = {
  x: 1,
  y: 2,
  z: {
    u: 3,
    w: 4,
  },
};

const SNAPSHOT = `\
// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[\`ex03 error diff-view 1\`] = \`
Object {
  "x": 1,
  "y": 2,
  "z": Object {
    "u": 3,
    "v": 4,
  },
}
\`;
`;

if (!fs.existsSync("test/__snapshots__")) {
  fs.mkdirSync("test/__snapshots__", { recursive: true });
}
fs.writeFileSync("test/__snapshots__/ex03.snap", Buffer.from(SNAPSHOT));

describe("ex03", () => {
  describe("error", () => {
    (updateSnapshot ? it.skip : it)("diff-view", () => {
      assert.throws(
        () => matchSnapshot(ACTUAL),
        (error: unknown) => {
          assert.ok(error instanceof SnapshotError);
          assert.ok(
            error
              .toString()
              .startsWith(
                "AssertionError [ERR_ASSERTION]: Snapshot not match (test/__snapshots__/ex03.snap)"
              )
          );
          return true;
        }
      );
    });
  });
});
