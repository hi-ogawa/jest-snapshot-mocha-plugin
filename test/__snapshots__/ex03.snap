// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ex03 error diff-view 1`] = `
Object {
  "x": 1,
  "y": 2,
  "z": Object {
    "u": 3,
    "v": 4,
  },
}
`;
